runtime/memprof.c

```diff
@@ -816,8 +816,8 @@ void caml_memprof_renew_minor_sample(void)
 }
 
 /* Called when exceeding the threshold for the next sample in the
-   minor heap, from the C code (the handling is different when called
-   from natively compiled OCaml code). */
+   minor heap. Execute allocation callbacks immediately unless if
+   called from C code or if asynchronous callbacks are masked. */
```

Comment:
Naive question: could we have a simpler system if we just considered that going to C masks asynchronous callbacks, to have a single notion instead of two? In particular:
- are FROM_C / FROM_CAML equivalent to assuming a particular mask level on the C side, or do they do other things?
- could we change `young_limit_c` to instead be `young_limit_for_mask_level_<foo>`, does that suggest a different code organization, and would we also get performance benefits when running masked OCaml code?

Reply:
Thinking more about this: even if we consider that `young_limit_c` is in fact `young_limit_no_large_callback`, a key point is that we have static knowledge that we are "within C code" so we statically know which limit to check. On the other hand, OCaml code running under a mask cannot (in general) know it without a dynamic check (or at least some offset computation to get the limit for the current mask level), which would be a performance regression in the limit-check hot path. This suggests that a conceptual unification may be possible, but that it does not magically improve the implementation.

(Note: we could have a check for the mask level at the entry of the cold path of Alloc_small, but that still probably is not worth it, as it increases code size for an optimization that would fire rarely.)

Reply:
* The C functions in the runtime are not considered as running in a mask: they do poll, they just decouple the allocations from the safe polling locations. In particular we want to control their polling location.

* As you explain it, this bug does not appear in OCaml code due to the dynamic nature of the mask and how it is implemented (whereby `young_limit` is always `young_limit_for_the_current_mask_level`).

That being said, I understand where the questions come from and I asked them myself when I was implementing them. To see how one arrives at this organisation, one has to start at the thought experiment at the end of the Motivation section of the PR description, where control on polling was given by a static annotation on OCaml functions. In that experiment, functions instructed to not poll would need to have the allocations replaced by ones that compare to the equivalent of `young_limit_c`, and programmers could insert polls by hand wher they see fit. Then, this would be the OCaml equivalent of what is currently done via the C API. So `young_limit_c` really means `young_limit_but_do_not_poll`. The conclusion of the thought experiment is that the purpose of a runtime flag compared to a static annotation is code re-use, so: - a way is devised to avoid this performance bug in the absence of static control (the same compiled code must perform allocations with and without mask), - existing polling locations in the C parts of the runtime must obey the mask (thus, running a C function must be different from running under a mask). This is how you get to the current organisation.